### PR TITLE
Make sdk `Handle` an opaque non-serializable type

### PR DESCRIPTION
**Pull request by blaxill**
Makes sdk `Handle` an opaque non-serializable type.

- `oak_runtime` keeps the Handle type alias to u64 as specified by the ABI. (This is used for example in `_raw` tests in `oak/examples/abitest/module/...` which are tests directly against the wasm which don't use the sdk)
- Tests updated to use opaque Handle where appropriate.  

#427

### CODE REVIEW

File: sdk/rust/oak/src/lib.rs [hunk 1]
```diff
@@ -37,7 +37,16 @@ use serial_test_derive::serial;
...
 ///
 /// These handles are used for all host function calls.
-pub type Handle = u64;
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct Handle {
+    id: u64,
+}
+
+impl Handle {
+    pub fn new (id: u64) -> Handle {
```
Comment (tiziano88) [inline sdk/rust/oak/src/lib.rs]: Is this code formatted correctly? There seems to be an extra whitespace.

File: sdk/rust/oak/src/lib.rs [hunk 2]
```diff
@@ -37,7 +37,16 @@ use serial_test_derive::serial;
 /// Handle used to identify read or write channel halves.
 ///
 /// These handles are used for all host function calls.
-pub type Handle = u64;
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct Handle {
```
Comment (tiziano88) [inline sdk/rust/oak/src/lib.rs]: FTR, an alternative here may be a "newtype", e.g. `pub struct Handle(u64)`, but it's probably worth keeping the proper struct (as you have done), in case we want to add more metadata, e.g. type of handle or "rights" to the underlying object.

See https://doc.rust-lang.org/rust-by-example/generics/new_types.html

File: sdk/rust/oak/src/lib.rs (hunk 1 above)
Comment (tiziano88) [inline sdk/rust/oak/src/lib.rs]: Maybe call this `from_raw` or similar? And also document that this should (usually) not be called from user code, but from serialisation and other internal libraries.

File: sdk/rust/oak/src/lib.rs [hunk 3]
```diff
@@ -191,10 +200,10 @@ pub fn channel_write(half: WriteHandle, buf: &[u8], handles: &[Handle]) -> OakSt
 /// On success, returns [`WriteHandle`] and a [`ReadHandle`] values for the
 /// write and read halves (respectively).
 pub fn channel_create() -> Result<(WriteHandle, ReadHandle), OakStatus> {
-    let mut write = WriteHandle { handle: 0 };
-    let mut read = ReadHandle { handle: 0 };
+    let mut write = WriteHandle { handle: Handle { id: 0 } };
```
Comment (tiziano88) [inline sdk/rust/oak/src/lib.rs]: Perhaps define a `Handle::invalid()` method (and corresponding `INVALID_HANDLE` const)?

File: examples/abitest/module2/rust/src/lib.rs [hunk 4]
```diff
@@ -37,15 +37,15 @@ pub fn main(in_handle: u64) -> i32 {
     oak::set_panic_hook();
 
     // We expect a single empty message holding a reply channel.
-    let in_channel = oak::ReadHandle { handle: in_handle };
+    let in_channel = oak::ReadHandle { handle: oak::Handle::new(in_handle) };
```
Comment (tiziano88) [inline examples/abitest/module2/rust/src/lib.rs]: Also change the type of the `in_handle` arg to `oak::Handle`, and perform the conversion in `oak_main`?

File: sdk/rust/oak_runtime/src/lib.rs [hunk 5]
```diff
@@ -29,6 +29,8 @@ use std::sync::{Arc, RwLock};
 
 pub mod proto;
 
+pub type Handle = u64;
```
Comment (tiziano88) [inline sdk/rust/oak_runtime/src/lib.rs]: Could this be made to use the one defined in the SDK? I think it would be fine (in fact, desirable) to have the runtime depend on the SDK crate (not the other way round of course).

**Comment by blaxill** [comment]
I think i've missed some of the tests. Working on it now.

File: sdk/rust/oak/src/lib.rs (hunk 1 above)
Comment (blaxill) [reply sdk/rust/oak/src/lib.rs]: I missed formatting originally, I think this was fixed in the formatting commit

File: sdk/rust/oak/src/lib.rs (hunk 2 above)
Comment (blaxill) [reply sdk/rust/oak/src/lib.rs]: Yeah, since the id is private, changing this would not require a large change either way.

File: sdk/rust/oak_runtime/src/lib.rs (hunk 5 above)
Comment (blaxill) [reply sdk/rust/oak_runtime/src/lib.rs]: I think this is backwards- `sdk/oak` is the safe interface that wraps WASM imported functions. `sdk/oak_runtime` is providing the WASM functions and so follows the ABI definitions (e.g. handle==u64).

I think the intention is for `oak_runtime` to move out of the `sdk` directory eventually, but currently it only powers the tests. @daviddrysdale does that sound right?

Comment (daviddrysdale) [reply sdk/rust/oak_runtime/src/lib.rs]: Yeah, that sounds about right.  

It would be nice to share definitions though, so it might make sense to have a little `oak_abi` crate that both the (Rust) runtime and the `oak` SDK can depend on. Looking at the current `oak_runtime/` code, I think it would currently just have:

 - the `oak::Handle` typedef
 - the `OakStatus` & `OakChannelStatus` enums (from protobuf definitions)
 - the `oak::wasm::INVALID_HANDLE` constant.

File: examples/abitest/module/rust/src/lib.rs [hunk 6]
```diff
@@ -173,22 +178,16 @@ fn invalid_raw_offset() -> *mut u64 {
...
+        let mut write = 0u64;
+        let mut read = 0u64;
         unsafe {
             expect_eq!(
                 OakStatus::ERR_INVALID_ARGS.value() as u32,
-                oak::wasm::channel_create(
-                    invalid_raw_offset() as *mut u64,
-                    &mut read.handle as *mut u64
-                )
+                oak::wasm::channel_create(invalid_raw_offset() as *mut u64, &mut read as *mut u64)
```
Comment (daviddrysdale) [inline examples/abitest/module/rust/src/lib.rs]: (pre-existing) Are the `as *mut u64` casts needed?  (Passim, with some `0` -> `0u64` conversions possibly needed.)

File: examples/abitest/module/rust/src/lib.rs [hunk 7]
```diff
@@ -479,35 +503,44 @@ impl FrontendNode {
...
             space.push(0x00);
             space
-                .write_u64::<byteorder::LittleEndian>(in_channel.handle)
+                .write_u64::<byteorder::LittleEndian>(in_channel)
                 .unwrap();
             space.push(0x00);
-            space
-                .write_u64::<byteorder::LittleEndian>(self.backend_in[0].handle)
-                .unwrap();
-            space.push(0x99); // deliberate nonsense status value
```
Comment (daviddrysdale) [inline examples/abitest/module/rust/src/lib.rs]: Why is this part of the test case dropped?

File: examples/abitest/module/rust/src/lib.rs [hunk 8]
```diff
@@ -573,21 +604,33 @@ impl FrontendNode {
...
-            OakStatus::OK,
-            oak::channel_read(in_channel, &mut buffer, &mut handles)
-        );
-        expect_eq!(3, buffer.len());
-        expect_eq!(0, handles.len());
+        let mut recv_handles = 0u32;
+        unsafe {
+            expect_eq!(
+                OakStatus::OK.value() as u32,
+                oak::wasm::channel_read(
```
Comment (daviddrysdale) [inline examples/abitest/module/rust/src/lib.rs]: Why the change from `oak::channel_read` to `oak::wasm::channel_read` ?

File: examples/abitest/module/rust/src/lib.rs [hunk 9]
```diff
@@ -743,14 +794,22 @@ impl FrontendNode {
     }
 
     fn test_node_create_raw(&self) -> std::io::Result<()> {
+        let mut out_channel = 0u64;
+        let mut in_channel = 0u64;
+
         unsafe {
+            oak::wasm::channel_create(&mut out_channel as *mut u64, &mut in_channel as *mut u64)
+                as i32;
+
```
Comment (daviddrysdale) [inline examples/abitest/module/rust/src/lib.rs]: Be nice to close the channel handles when you're done.

File: examples/abitest/module2/rust/src/lib.rs [hunk 10]
```diff
@@ -37,15 +37,17 @@ pub fn main(in_handle: u64) -> i32 {
...
+    };
     let mut buf = Vec::<u8>::with_capacity(2);
     let mut handles = Vec::with_capacity(2);
     oak::channel_read(in_channel, &mut buf, &mut handles);
     if handles.len() != 1 {
         return oak::OakStatus::ERR_INTERNAL.value();
     }
     let out_handle = handles[0];
-    info!("backend node: in={}, out={}", in_handle, out_handle);
+    info!("backend node: in={}, out={:?}", in_handle, out_handle);
```
Comment (daviddrysdale) [inline examples/abitest/module2/rust/src/lib.rs]: Maybe implement `Display` for `oak::Handle` ?

File: sdk/rust/oak/src/grpc/mod.rs [hunk 11]
```diff
@@ -144,7 +144,7 @@ pub trait OakNode {
 /// [`GrpcRequest`]: crate::proto::grpc_encap::GrpcRequest
 pub fn event_loop<T: OakNode>(mut node: T, grpc_in_handle: ReadHandle) -> i32 {
     info!("start event loop for node with handle {:?}", grpc_in_handle);
-    if grpc_in_handle.handle == wasm::INVALID_HANDLE {
+    if grpc_in_handle.handle.id == wasm::INVALID_HANDLE {
```
Comment (daviddrysdale) [inline sdk/rust/oak/src/grpc/mod.rs]: Maybe there should be a `Handle::is_invalid()` method (or somesuch) so that things outside of the `oak` crate can check this too?

File: sdk/rust/oak_runtime/src/lib.rs (hunk 5 above)
Comment (tiziano88) [reply sdk/rust/oak_runtime/src/lib.rs]: That's what I meant, except I think it's fine for the runtime crate to depend on the SDK crate; anyways it's fine to add a shared `oak_abi` crate that both depends on if you prefer.

File: examples/abitest/module2/rust/src/lib.rs (hunk 10 above)
Comment (tiziano88) [reply examples/abitest/module2/rust/src/lib.rs]: I would prefer `Debug`, since this should really not be needed for anything other than debugging (for similar reasons why we are trying to prevent code from naively serialising it too).

File: examples/abitest/module/rust/src/lib.rs (hunk 7 above)
Comment (blaxill) [reply examples/abitest/module/rust/src/lib.rs]: `backend_in` contains `oak::Handle`s which now cannot be serialised, `sdk/rust/oak` doesn't export a method for getting their u64 value. I think this was original intended as a erroneous value that should fail? I could:

- replace it with a different erroneous value
- use `oak_runtime::Handle`s for backend_in

But it seems that making `oak::Handle` opaque and non-serialisable has removed the need for the current definition (it makes it impossible!).

File: examples/abitest/module/rust/src/lib.rs (hunk 8 above)
Comment (blaxill) [reply examples/abitest/module/rust/src/lib.rs]: I was trying to switch to exclusively using `oak::wasm` functions in `_raw` tests and `oak::` in others. The new opaque type requires that non-raw functions use the opaque `oak::Handle` type as the handles can't be unwrapped, but the `_raw` functions can mix both I guess. It seemed therefore more consistent to only use `oak::wasm` functions in the `_raw` functions.

File: examples/abitest/module2/rust/src/lib.rs (hunk 10 above)
Comment (blaxill) [reply examples/abitest/module2/rust/src/lib.rs]: I'm open to this, but as the type is an opaque wrapper there is some logical consistency to requiring a debugging-print to see inside

File: examples/abitest/module/rust/src/lib.rs (hunk 6 above)
Comment (blaxill) [reply examples/abitest/module/rust/src/lib.rs]: Some are needed for the ref to ptr part of the cast, but others can be removed. I think abstracting the following somewhere makes sense since this is used in a number of "raw" tests. 

~~~rust
fn channel_create_raw(&self) -> (u64,u64) {
    let mut w = 0u64;
    let mut r  = 0u64;
    unsafe { oak::wasm::channel_create(&mut w as *mut u64, &mut r as *mut u64) as i32; }
    (w,r)
}
~~~

File: examples/abitest/module/rust/src/lib.rs (hunk 7 above)
Comment (blaxill) [reply examples/abitest/module/rust/src/lib.rs]: (The other handles in `write_u64` calls above have been changed to the `oak_runtime::Handle` which is still transparent)

File: examples/abitest/module2/rust/src/lib.rs (hunk 10 above)
Comment (daviddrysdale) [reply examples/abitest/module2/rust/src/lib.rs]: Yeah, `Debug`-only does makes sense.

File: sdk/rust/oak_runtime/src/lib.rs (hunk 5 above)
Comment (daviddrysdale) [reply sdk/rust/oak_runtime/src/lib.rs]: The runtime crate depending on the SDK crate seems like a layering inversion &ndash; it should theoretically be possible for a masochistic Node author to write raw Wasm code and therefore never even install the SDK crate(s).

File: examples/abitest/module/rust/src/lib.rs (hunk 7 above)
Comment (daviddrysdale) [reply examples/abitest/module/rust/src/lib.rs]: IIRC the test case was intended to get a variety of different channel read status values back, so if there were another channel available that would return `NOT_READY` then I think the same ground would be covered.

File: examples/abitest/module/rust/src/lib.rs (hunk 8 above)
Comment (daviddrysdale) [reply examples/abitest/module/rust/src/lib.rs]: OK, that makes sense.

File: sdk/rust/oak_runtime/src/lib.rs (hunk 5 above)
Comment (tiziano88) [reply sdk/rust/oak_runtime/src/lib.rs]: Not sure I follow; I am just saying that we can consider the sdk crate as the "public interface" of the runtime crate, so it seems fine to have the runtime depend on the sdk crate, and have all "public" types in the sdk crate but also be able to use them from the runtime crate internally. I don't see any layering violation there; no one externally would ever use the runtime crate directly, right?

Or, again, as @daviddrysdale suggested, it's also fine to have a shared crate that both sdk and runtime depend on.

Comment (daviddrysdale) [reply sdk/rust/oak_runtime/src/lib.rs]: I don't think the sdk crate is the public interface of the runtime &ndash; the collection of host functions that make up the ABI is the public interface of the runtime.  The SDK is then a collection of libraries that make it (much) easier for Rust code to use that ABI; if a Node were written in (say) C++ then that SDK wouldn't be available.

To put it in analogy terms with UNIX:
 - Oak runtime ≈ UNIX kernel
 - Oak host functions ABI ≈ UNIX system call interface
 - Oak SDK ≈ libc (and hopefully other user-space libraries too).

It would be odd for the Linux kernel to depend on a header file from glibc; if nothing else, it would prevent the kernel being used with (say) [musl](https://www.musl-libc.org/).

Comment (tiziano88) [reply sdk/rust/oak_runtime/src/lib.rs]: Ah got it, thanks for clarifying, I also re-read your previous message more carefully and that does make sense.

Yes, if we want to make the sdk crate optional for node authors (at least in principle), then a minimal ABI crate is the way to go. My point was just that technically it would be fine for the runtime crate to depend on the sdk crate, but not the other way round. Having the ABI one is even better for layering / separation purposes, I agree.

Comment (daviddrysdale) [reply sdk/rust/oak_runtime/src/lib.rs]: #447 is working towards having the Rust runtime not under SDK, and depending on an `oak_abi` crate for ABI type/constant definitions.

File: examples/chat/module_1/rust/src/lib.rs [hunk 12]
```diff
@@ -35,7 +35,9 @@ pub fn main(in_handle: u64) -> i32 {
     let mut rsp_writers: Vec<oak::grpc::ChannelResponseWriter> = Vec::with_capacity(1);
     loop {
         // Wait for something on our single input channel.
-        let in_channel = oak::ReadHandle { handle: in_handle };
+        let in_channel = oak::ReadHandle {
```
Comment (tiziano88) [inline examples/chat/module_1/rust/src/lib.rs]: Perhaps move this before the start of the loop.

File: examples/abitest/module_0/rust/src/lib.rs [hunk 13]
```diff
@@ -165,30 +170,32 @@ fn status_convert<T>(result: Result<T, OakStatus>) -> std::io::Result<T> {
 }
 
 // Return an offset value that is beyond the bounds of available linear memory.
-fn invalid_raw_offset() -> *mut u64 {
+unsafe fn invalid_raw_offset() -> *mut u64 {
```
Comment (daviddrysdale) [inline examples/abitest/module_0/rust/src/lib.rs]: BTW, why didn't this need to be marked as `unsafe` before?

File: sdk/rust/oak/src/lib.rs [hunk 14]
```diff
@@ -37,7 +37,30 @@ use serial_test_derive::serial;
...
+    pub fn from_raw(id: u64) -> Handle {
+        Handle { id }
+    }
+
+    /// Check this handle is valid.
+    pub fn is_valid(self) -> bool {
+        self.id != wasm::INVALID_HANDLE
+    }
+
+    /// Returns an intentionally invalid handle. Intended for debugging purposes.
```
Comment (tiziano88) [inline sdk/rust/oak/src/lib.rs]: Why for debugging purposes? It is already used in `channel_create`, right?

File: sdk/rust/oak_runtime/src/lib.rs (hunk 5 above)
Comment (tiziano88) [reply sdk/rust/oak_runtime/src/lib.rs]: So is the plan to remove this in favour of the upcoming shared definition? If so, perhaps add a `TODO(#447)` for the time being?

Or are you suggesting keeping both `oak::Handle` (from the SDK) and `oak_abi::Handle` around eventually?

File: examples/abitest/module_0/rust/src/lib.rs [hunk 15]
```diff
@@ -479,35 +497,40 @@ impl FrontendNode {
...
             space.push(0x00);
             space
-                .write_u64::<byteorder::LittleEndian>(in_channel.handle)
+                .write_u64::<byteorder::LittleEndian>(in_channel)
                 .unwrap();
             space.push(0x00);
-            space
-                .write_u64::<byteorder::LittleEndian>(self.backend_in[0].handle)
-                .unwrap();
-            space.push(0x99); // deliberate nonsense status value
```
Comment (daviddrysdale) [inline examples/abitest/module_0/rust/src/lib.rs]: Would still be nice not to lose the `NOT_READY` subcase.

File: examples/abitest/module_1/rust/src/lib.rs [hunk 16]
```diff
@@ -33,19 +33,20 @@ pub extern "C" fn oak_main(handle: u64) -> i32 {
     }
 }
 pub fn main(in_handle: u64) -> i32 {
+    let in_channel = oak::ReadHandle {
+        handle: oak::Handle::from_raw(in_handle),
+    };
```
Comment (daviddrysdale) [inline examples/abitest/module_1/rust/src/lib.rs]: nit: this should move down by 2 lines (initializing logging and setting the panic hook should be the first things done)

File: examples/abitest/module_0/rust/src/lib.rs (hunk 13 above)
Comment (blaxill) [reply examples/abitest/module_0/rust/src/lib.rs]: It actually doesn't need to be marked unsafe, would you prefer this reverted? Only dereferencing it is unsafe, but I thought this made the intention more clear.

File: examples/abitest/module_0/rust/src/lib.rs (hunk 15 above)
Comment (blaxill) [reply examples/abitest/module_0/rust/src/lib.rs]: I've added a TODO, so this PR isn't held up longer

File: examples/abitest/module_0/rust/src/lib.rs (hunk 13 above)
Comment (daviddrysdale) [reply examples/abitest/module_0/rust/src/lib.rs]: Fine as-is, I just wondered if something had changed to make it necessary that I couldn't see.
